src/diffpy/utils/diffraction_objects.py

```diff
@@ -248,15 +248,45 @@ def _set_array_from_range(self, begin, end, step_size=None, n_steps=None):
             array = np.linspace(begin, end, n_steps)
         return array
 
-    def get_angle_index(self, angle):
-        count = 0
-        for i, target in enumerate(self.angles):
-            if angle == target:
-                return i
-            else:
-                count += 1
-        if count >= len(self.angles):
-            raise IndexError(f"WARNING: no angle {angle} found in angles list")
+    def get_array_index(self, value, xtype=None):
+        """
+        returns the index of the closest value in the array associated with the specified xtype
+
+        Parameters
+        ----------
+        xtype str
+            the xtype used to access the array
+        value float
+            the target value to search for
+
+        Returns
+        -------
+        the index of the value in the array
+        """
+
+        if xtype is None:
+            xtype = self.input_xtype
+        if self.on_xtype(xtype) is None or len(self.on_xtype(xtype)[0]) == 0:
+            raise ValueError(
+                f"The '{xtype}' array is empty. " "Please ensure it is initialized and the correct xtype is used."
+            )
+        array = self.on_xtype(xtype)[0]
+        i = (np.abs(array - value)).argmin()
+        nearest_value = np.abs(array[i] - value)
+        distance = min(np.abs(value - array.min()), np.abs(value - array.max()))
+        threshold = 0.5 * (array.max() - array.min())
+
+        if nearest_value != 0 and (array.min() <= value <= array.max() or distance <= threshold):
```

Comment:
I am rethinking all this.  I wonder if we should just return i and be done with it.  Raise an error if the array is empty as you do, but then just return i.  what do you think?  Then it is up to the user to make sure the result makes sense.